api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+	txn txn
+	err error
+}
+
+func (s *Service) wait(tx txn) chan waitResult {
+	s.waitingLck.Lock()
+	defer s.waitingLck.Unlock()
+
+	waitCh, found := s.waiting[tx.ID]
+	if found {
+		return waitCh
+	}
+
+	ch := make(chan waitResult)
+	s.waiting[tx.ID] = ch
+
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), s.config.MessageWaitTimeout)
+
+		client, closeClient, err := s.clientBuilder(ctx)
+		defer func() {
+			closeClient()
+			close(ch)
+			cancel()
+		}()
+		if err != nil {
+			ch <- waitResult{err: fmt.Errorf("creating lotus client: %v", err)}
+			return
+		}
+
+		lastestMsgCid, err := tx.latestMsgCid()
+		if err != nil {
+			tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+			tx.FailureMsg = fmt.Sprintf("getting latest message cid from txn: %v", err)
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			return
+		}
+		c, err := cid.Decode(lastestMsgCid.Cid)
+		if err != nil {
+			tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+			tx.FailureMsg = fmt.Sprintf("decoding latest message cid for txn: %v", err)
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			return
+		}
+
+		tx.Waiting = true
+		if err := s.updateTxn(ctx, tx); err != nil {
+			ch <- waitResult{err: err}
+			return
+		}
+
+		res, err := client.StateWaitMsg(ctx, c, s.config.MessageConfidence)
+		tx.Waiting = false
+		if err != nil {
+			// If for some reason the lotus node doesn't know about the cid, consider that a final error.
+			if strings.Contains(err.Error(), "block not found") {
+				tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+				tx.FailureMsg = err.Error()
+				tx.UpdatedAt = time.Now()
+				log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			}
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			if tx.MessageState == pb.MessageState_MESSAGE_STATE_FAILED {
+				ch <- waitResult{txn: tx}
+			} else {
+				ch <- waitResult{err: fmt.Errorf("calling StateWaitMsg: %v", err)}
+			}
+			return
+		}
+
+		if res.Receipt.ExitCode.IsError() {
+			tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+			tx.FailureMsg = fmt.Sprintf("error exit code: %v", res.Receipt.ExitCode.Error())
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			if res.Receipt.ExitCode.IsSendFailure() {
+				log.Errorf("received exit code send failure: %s", res.Receipt.ExitCode.String())
+			} else {
+				log.Infof("received exit code error: %s", res.Receipt.ExitCode.String())
+			}
+			return
+		}
```

Comment:
If the receipt data we get from lotus indicates that the exit code is an error exit code, we consider this a final error. We do some extra logging to get a feel for if the exit code is a "send failure" (which would be strange and alarming) or some other more normal failing exit code.